src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -251,69 +226,134 @@ protected virtual string GenerateUniqueId()
                                 .WriteStartMethodInvocation(_tagHelperContext.FormatInvalidIndexerAssignmentMethodName)
                                 .WriteStringLiteral(attributeName)
                                 .WriteParameterSeparator()
-                                .WriteStringLiteral(tagHelperTypeName)
+                                .WriteStringLiteral(tagHelperDescriptor.TypeName)
                                 .WriteParameterSeparator()
                                 .WriteStringLiteral(attributeDescriptor.PropertyName)
                                 .WriteEndMethodInvocation(endLine: false)   // End of method call
                                 .WriteEndMethodInvocation(endLine: true);   // End of new expression / throw statement
                         }
                     }
-
-                    var dictionaryKey = attributeName.Substring(attributeDescriptor.Name.Length);
-                    valueAccessor = string.Format(
-                        CultureInfo.InvariantCulture,
-                        "{0}.{1}[\"{2}\"]",
-                        tagHelperVariableName,
-                        attributeDescriptor.PropertyName,
-                        dictionaryKey);
                 }
+            }
+        }
+
+        private void RenderAttributes(
+            IList<KeyValuePair<string, Chunk>> chunkAttributes,
+            IEnumerable<TagHelperDescriptor> tagHelperDescriptors)
+        {
+            // First attribute wins, even if there are duplicates.
```

Comment:
Got it now. Leave out the comment because it makes sense only where the `HashSet` is checked. Variable name is :ok: though "Names" part doesn't add much.